README.md
# flickrPHP
PHP Script which loads a certain flickr Photoset / Gallery in a Multicolumn Site. Uses Fancybox JS to show large versions of the image onclick. Fallback for multiple image sizes. Local Caching with JSON file.

<a href="http://beta.nx-designs.ch/flickr/phpflickr/flickrPHP_large.php" target="_blank">Live Demo with more then 250 Images (Cached)</a>

## Requirements
- Server running PHP Version 7.x
- flickr Account (to obtain an API Key)
- flickr API Key (<a href="https://www.flickr.com/services/apps/create/apply" target="_blank">get one here</a>)

## Used JS Frameworks
- jQuery 3.x
- yootheme uikit 3 (beta 22)
- Fancybox 3

## Detail
- Responsive Grid System for Galleryview
- API-Key is stored as PHP Var
  - If showinfo is turned off Users cant see your flickr API Key
- Lightbox with Fancybox 3
- Scriptcheck for different Imagesizes (Lightbox) Best will be used
  - Large 2048
  - Large 1600
  - Large 1024
  - Fallback: Original
- Debug Output

## Settings
### showdebug
#### true / false
Displays Debug Informations.<br>
Informations will be placed direct on the Page and into the console.<br>
<a href="http://beta.nx-designs.ch/flickr/phpflickr/flickrPHP_debug.php" target="_blank">Demo with Debug Mode on</a>
### showinfo
#### true / false
Displays Connection Informations in a grey Box on Top of the Gallery.
### API Key
#### String
You need your own API Key to use the flickr API. You can get your flickr API Key <a href="https://www.flickr.com/services/apps/create/apply" target="_blank">here</a> for free.
### Photoset ID
#### String
The ID of the desired Gallery you would like to show. You can find the Gallery / Photoset ID directly on flickr. Just copy the ID from the URL: .../149621562@N02/albums/<b>72157682900004326</b>. For example 72157682900004326 is the Gallery ID you can use.

## Cache
To limit the flickr connections to a minimum the script creates a local cache for your photosets & images. The flickr responses are stored in .JSON Files. The Files can be found under cache/"galleryID"/...
Because of the Caching procedure it could happen that the first Pageload can take a long time. Please be patient and if you receive an error message just reload the page. The script continues and load the necessary informations from flickr.
The Script demands from flickr the gallery Informations and the imagesize informations for each photo.
Once the Cache is created the Site loads quick depending on the amount of images you have in your gallery. In our example you can see a gallery with more then 250 Pictures.
  


changelog.md
## version 1.1 updates
a few new options to set:<br>
#### showheader <br>						
displays the black header on top<br><br>
#### showdebug<br>							
shows debug informations on page<br><br>
#### showinfo<br>
shows grey infobox with API Key, Cache State, Photoset Informations on top<br><br>
#### showtitle<br>
Display Photoset Title<br><br>
#### loadjquery<br>
should jQuery (3) been loaded by this script?<br><br>
#### framework<br>
uikit || Bootstrap<br><br>
#### loadfrontendframework<br>
should grid-frameworks be loaded by this script?<br><br>
#### loadfancyBoxScript<br>
load fancyBox 3 Script for PopUp?<br><br>
#### enablepopups<br>
select if popups with fancybox should be enabled<br><br>
#### columns<br>
choose desired amount of columns (for uikit = 1,2,3,4,5,6 for bootstrap = 1,2,3,4,6,12)<br><br>
#### thumbnailsize<br>
s,q,t,m,n,z,c,b,h,k,o (depends on available image quality on flickr for this set)<br><br>					


flickrPHP.php
<?php
/*

	name:			flickr PHP Script
	version:		1.1
	Description:	flickr PHP Script
	creator:		nx-designs | Marco Rensch
	web:			http://www.nx-designs.ch
	repository:		https://github.com/marcorensch/flickrPHP
	type:			opensource
	license:		GNU General Public License version 3 (https://opensource.org/licenses/gpl-3.0.html)
	
	important:		For this script fancyBox 3 is used as default PopUp Gallery. 
					Please read license informations for commercial usage
					here: http://fancyapps.com/fancybox/3/#license.
	

*/
	
	// Global Parameters & Photoset Informations
	$showheader = true;								// displays the black header on top 				 	
	$showdebug = false;								// shows debug informations on page						
	$showinfo = true;								// shows grey infobox with API Key, Cache State, Photoset Informations on top
	$showtitle = true;								// Display Photoset Title
	$loadjquery = true;								// should jQuery (3) been loaded by this script?
	$framework = 'Bootstrap'; 						// uikit || Bootstrap
	$loadfrontendframework = true; 					// should grid-frameworks be loaded by this script?
	$loadfancyBoxScript = true;						// load fancyBox 3 Script for PopUp?
	$grouping = 'false';
	$enablepopups = true; 							// select if popups with fancybox should be enabled
	$columns = 12;									// choose desired amount of columns (for uikit = 1,2,3,4,5,6 for bootstrap = 1,2,3,4,6,12)
	$thumbnailsize = 'q';							// s,q,t,m,n,z,c,b,h,k,o (depends on available image quality on flickr for this set)
	$apikey = 'a1200b7adcda28d652f1583e57a3bd85';	// Enter here your API Key
	$photosetID = '72157680841776090'; 			// Demo Gallery Golf (250+ Images)
	// $photosetID = '72157682900004326'; 				// Demo Gallery Fox (6 Images)

	$photoset_url = "";
	$photoset_JSON_url = "";
	$photoset_length = 0;
	$photoset_title = "";
	
	include('helper.php');
	include('setconfig.php');
?>
<!doctype html>
<html>
<head>
<meta charset="UTF-8">
<title>PHP flickr Requests</title>
<?php if($loadjquery): ?>
<!-- jQuery -->
	<script
			  src="https://code.jquery.com/jquery-3.2.1.min.js"
			  integrity="sha256-hwg4gsxgFZhOsEEamdOYGBf13FyQuiTwlAQgxVSNgt4="
			  crossorigin="anonymous">
	</script>
<!-- End jQuery -->
<?php endif; ?>
<?php if($loadfrontendframework): ?>
		<!-- <?php echo $framework; ?> is in use -->
		<link rel="stylesheet" type="text/css" <?php echo $stylesheetlink; ?> />
		<script type="text/javascript" <?php echo $frameworklink; ?>></script>
		<!-- End <?php echo $framework; ?> -->
<?php endif; ?>

<!-- FancyBox -->
<?php if($loadfancyBoxScript): ?>
	<link rel="stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.0.47/jquery.fancybox.min.css" />
	<script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.0.47/jquery.fancybox.min.js"></script>
<?php endif; ?>
<!-- End FancyBox -->

<!-- General Stylesheet -->
	<link rel="stylesheet" type="text/css" href="css/main.css" />
<!-- End General Styles -->


</head>

<body>
	<?php if($showheader): ?>
	<div id="header">
		<h1 id='nx-sitetitle'>flickr PHP Script</h1>
		<h3 id="nx-titleby">by nx-designs</h3>
	</div>
	<?php endif; ?>
	<div id="main">
		<?php if($showinfo) : ?>
		<div class="uk-container-large uk-padding container-fluid">
			<div id="nx-connectionInfo" class="<?php echo $infoboxclass; ?>">
				<?php echo 'API Key: '.$apikey."<br>\n"; ?>
				<?php echo 'Photoset ID: '.$photosetID."<br>\n"; ?>
				<?php if($photoset_url != ''){
						echo '<span class="nx-alertmsg">Set was not loaded...<br>Informations Downloaded from flickr<br>Complete Request URL: <a href="'.$photoset_url.'" target="_blank">'.$photoset_url."</a></span><br>";
					}else {
						echo '<span>Set was loaded, cached version can be found here: <a href="'.$photoset_JSON_url.'" target="_blank">'.$photoset_JSON_url.'</a></span><br>';
					} ?>
				<?php echo 'The Gallery contains <b> '.$photoset_length.' </b>Elements. <br>';?>
			</div>
		</div>
		<?php endif; ?>
		<?php if($showtitle): ?>
		<div class="uk-container-large uk-padding container-fluid">
			<h1 style="text-align:center;"><?php echo $photoset_title; ?></h1>
			<hr>
		</div>
		<?php endif; ?>
		<div class="uk-container-large uk-padding container-fluid">
			<div id="nx-gallery" class="<?php echo $outerdiv; ?>" uk-grid>
			<?php
				// Get Images & Build Thumbnails
				for ($i = 0; $i < $photoset_length; $i++) {
					$imgID = $photoset_array[$i]->id;
					$farmID = $photoset_array[$i]->farm;
					$secret = $photoset_array[$i]->secret;
					$serverID = $photoset_array[$i]->server;
					$imgtitle = $photoset_array[$i]->title;

					$imglargeurl = checkCache(2,$imgID,$apikey);
					if($imglargeurl == ''){
						$imglargeurl = "https://farm'.$farmID.'.staticflickr.com/'.$serverID.'/'.$imgID.'_'.$secret.'_q.jpg";
						debug_to_console('Attention no large image available please check debug');
					}
					
					echo '<div class="nx-block '.$gridelement.'" id="nx-img_'.$i.'">';
					echo '<a '.$lightbox.' data-caption="'.$imgtitle.'" href="'.$imglargeurl.'" target="_blank"><img title="'.$imgtitle.'" src="https://farm'.$farmID.'.staticflickr.com/'.$serverID.'/'.$imgID.'_'.$secret.'_'.$thumbnailsize.'.jpg" width="100%" /></a>';
					echo '</div>';
				}
			?>
			</div>
		</div>
	</div>
</body>
</html>

setconfig.php
<?php
if($loadfrontendframework):
	switch ($framework){
		case 'uikit':
			$frameworklink = 'src="https://cdnjs.cloudflare.com/ajax/libs/uikit/3.0.0-beta.22/js/uikit.min.js"';
			$stylesheetlink = 'href="https://cdnjs.cloudflare.com/ajax/libs/uikit/3.0.0-beta.22/css/uikit.min.css"';
			$outerdiv = 'uk-child-width-1-'.$columns;
			$gridelement = '';
			$infoboxclass = 'uk-width-1-1';
			break;
		case 'Bootstrap':
			switch ($columns){
				case 1: $bswidth = 12; break;
				case 2: $bswidth = 6; break;
				case 3: $bswidth = 4; break;
				case 4: $bswidth = 3; break;
				case 6: $bswidth = 2; break;
				case 12: $bswidth = 1; break;
			}
			$frameworklink = 'src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"';
			$stylesheetlink = 'href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous"';
			$outerdiv = 'row';
			$gridelement = 'col-sm-'.$bswidth;
			$infoboxclass = 'col-sm-12';
			break;
		default :
			$frameworklink = '';
			$stylesheetlink = '';
			$outerdiv = '';
			$gridelement = '';
			break;
	}
endif;
// Prepare for PopUp
if($enablepopups){
	switch($grouping){
		case true :
			$lightbox = 'data-fancybox="group"';
			break;
		case false:
			$lightbox = 'data-fancybox=""';
			break;
	}
}else{
		$lightbox = '';
	}
?>

helper.php
<?php
// Helper File

// Function for Console Output
	
function debug_to_console( $data ) {
	$output = $data;
	if ( is_array( $output ) )
	$output = implode( ',', $output);

	echo "<script>console.log( 'Debug Information: " . $output . "' );</script>";
}



if($apikey == ''){
	$noapikeywarn = '<div class="nx-block nx-center nx-warning"><b>WARNING</b><br>Your API Key is not set. Please enter your Flickr API Key in the settings.<br>You can get an API Key <a href="https://www.flickr.com/services/api/misc.api_keys.html" target="_blank">here</a></div>';
	echo $noapikeywarn;
	alert($noapikeywarn);	
}
if($photosetID == ''){
	$nosetidwarn = '<div class="nx-block nx-center nx-warning"><b>WARNING</b><br>There is no Photoset ID given. Please enter a Flickr Photoset ID in the settings.</div>';
	echo $nosetidwarn;
	alert($nosetidwarn);	
}

function setvar($type,$url){
	// use global var's
	global $photoset_array;
	global $photoset_JSON_url;
	global $photoset_length;
	global $photoset_title;

	$response = json_decode(file_get_contents($url));
	switch($type){
		case 1:		// Get Gallery Informations (photoarray,length, title)
			$photoset_JSON_url = $url;
			$photoset_array = $response->photoset->photo;
			$photoset_length = $response->photoset->total;
			$photoset_title = $response->photoset->title;
			break;
		case 2:		// nothing to do if we start the function with the JSON for imagesizes
			return $response;
			break;
	}
}

checkCache(1,$photosetID,$apikey);

function checkCache($type,$ID,$apikey){
	global $showdebug;
	global $photosetID;
	switch($type){
		case 1:		// set $file path if we check for photoset informations
			$file = "cache/$photosetID/photoset_$ID.json";
			break;
		case 2:		// set $file path if we check for image informations
			$file = "cache/$photosetID/img_$ID.json";
			break;
	}
	if (file_exists($file)) {
		if($showdebug){
			debug_to_console( "The File $file exists already");
		}
		$response = setvar($type,$file);
	} else {
		if($showdebug){
			debug_to_console( "The File $file does not exist");
		}
		// if file does not exists get the informations and create it
		$filecreated = createCache($type,$ID,$apikey);
		$response = setvar($type, $filecreated);
		if($showdebug){print_r($response);}
	}
	if($type == 2){
		// if we use the function for image informations get now the large Image URL from the JSON
		$imglargeurl = getZoomedIMG($response);
		return $imglargeurl;	
	}
}
// creates the JSON Files into the chache folder
function createCache($type,$ID,$apikey){
	global $photosetID;
	global $showdebug;

	switch ($type){
		case 1:		// create folder & file for the photoset informations
			global $photoset_url;
			$photoset_url = "https://api.flickr.com/services/rest/?method=flickr.photosets.getPhotos&api_key=$apikey&photoset_id=$ID&format=json&nojsoncallback=1";
			$photoset = json_decode(file_get_contents($photoset_url),true);
			if (!file_exists('cache/'.$photosetID)) {
				mkdir('cache/'.$photosetID, 0777, true);
			}
			$fp = fopen('cache/'.$photosetID.'/photoset_'.$ID.'.json', 'w');
			fwrite($fp, json_encode($photoset));
			fclose($fp);
			$filecreated = "cache/$photosetID/photoset_$ID.json"; // "cache/123456789/photoset_987654321.json"
			break;
		case 2:		// create file for image informations (each image has its own .JSON File)
			$url = "https://api.flickr.com/services/rest/?method=flickr.photos.getSizes&api_key=$apikey&photo_id=$ID&format=json&nojsoncallback=1";
			$sizesArray = json_decode(file_get_contents($url),true);
			$fp = fopen('cache/'.$photosetID.'/img_'.$ID.'.json', 'w');
			fwrite($fp, json_encode($sizesArray));
			fclose($fp);
			$filecreated = "cache/$photosetID/img_$ID.json"; // "cache/123456789/img_987654321.json"
			break;
	}
	if($showdebug){debug_to_console( "The File $filecreated was successfully created");}
	return($filecreated);
}

function getZoomedIMG($imgDetails){
	global $showdebug;
	if($showdebug){
		echo '<div class="uk-width-1-1"><br><hr><h3>Image Details</h3>';
		print_r ($imgDetails->sizes->size);
		echo '<br></div>';
		echo '<div class="uk-width-1-1"><h4>Maximal Available Size</h4>';
	}
	if (in_array("Large 2048", array_column($imgDetails->sizes->size, 'label'))) {
		if($showdebug){echo 'Large 2048 is available<br>';}

		$key = array_search('Large 2048', array_column($imgDetails->sizes->size, 'label'));
		if($showdebug){echo 'Large 2048 is available on position '.$key.'<br></div>';}

		$imglargeurl = $imgDetails->sizes->size[$key]->source;

		if($showdebug){echo '<div class="uk-width-1-1"><h4>Large Image Link</h4><a href="'.$imglargeurl.'" target="_blank">'.$imglargeurl.'</a></div>';}

	}elseif(in_array("Large 1600", array_column($imgDetails->sizes->size, 'label'))){
		if($showdebug){echo 'Large 1600 is available<br>';}

		$key = array_search('Large 1600', array_column($imgDetails->sizes->size, 'label'));
		if($showdebug){echo 'Large 1600 is available on position '.$key.'<br></div>';}

		$imglargeurl = $imgDetails->sizes->size[$key]->source;

		if($showdebug){echo '<div class="uk-width-1-1"><h4>Large Image Link</h4><a href="'.$imglargeurl.'" target="_blank">'.$imglargeurl.'</a></div>';}
	}elseif(in_array("Large 1024", array_column($imgDetails->sizes->size, 'label'))){
		if($showdebug){echo 'Large 1024 is available<br>';}

		$key = array_search('Large 1024', array_column($imgDetails->sizes->size, 'label'));
		if($showdebug){echo 'Large 1024 is available on position '.$key.'<br></div>';}

		$imglargeurl = $imgDetails->sizes->size[$key]->source;

		if($showdebug){echo '<div class="uk-width-1-1"><h4>Large Image Link</h4><a href="'.$imglargeurl.'" target="_blank">'.$imglargeurl.'</a></div>';}
	}elseif(in_array("Original", array_column($imgDetails->sizes->size, 'label'))){
		if($showdebug){echo 'Original is available<br>';}

		$key = array_search('Original', array_column($imgDetails->sizes->size, 'label'));
		if($showdebug){echo 'Original is available on position '.$key.'<br></div>';}

		$imglargeurl = $imgDetails->sizes->size[$key]->source;

		if($showdebug){echo '<div class="uk-width-1-1"><h4>Large Image Link</h4><a href="'.$imglargeurl.'" target="_blank">'.$imglargeurl.'</a></div>';}
	}else{
		$imglargeurl = '';
		if($showdebug){echo '<b>There where no large images available.</b></div>';}
		if($showdebug){debug_to_console('There where no large images available');}
	}
	return $imglargeurl;
}

?>